internal/perf/perf.go

```diff
@@ -603,53 +644,88 @@ func (pr *perfRunner) runLoop(tc TestCase) error {
 				log.Errorf("Error retreiving histogram: %s", histErr)
 			}
 
-			// Wait until the rate limiter allows us to proceed
-			limiter.Wait(pr.ctx)
-
 			startTime := time.Now()
-			trackingID, err := tc.RunOnce()
+			trackingIDs := make([]string, 0)
 
-			if err != nil {
-				if pr.cfg.DelinquentAction == conf.DelinquentActionExit.String() {
-					return err
+			for actionsCompleted = 0; actionsCompleted < tc.ActionsPerLoop(); actionsCompleted++ {
+				// Wait until the rate limiter allows us to proceed
+				limiter.Wait(pr.ctx)
+
+				if pr.allActionsComplete() {
+					break
+				}
+
+				trackingID, err := tc.RunOnce()
+
+				if err != nil {
+					if pr.cfg.DelinquentAction == conf.DelinquentActionExit.String() {
+						return err
+					} else {
+						log.Errorf("Worker %d error running job (logging but continuing): %s", workerID, err)
+						err = nil
+						continue
+					}
 				} else {
-					log.Errorf("Worker %d error running job (logging but continuing): %s", workerID, err)
-					err = nil
-					continue
+					trackingIDs = append(trackingIDs, trackingID)
+					totalActionsCounter.Inc()
 				}
-			} else {
-				totalActionsCounter.Inc()
-			}
 
-			pr.markTestInFlight(tc, trackingID)
-			log.Infof("%d --> %s Sent %s: %s", workerID, testName, idType, trackingID)
+				pr.markTestInFlight(tc, trackingID)
+				log.Infof("%d --> %s Sent %s: %s", workerID, testName, idType, trackingID)
+			}
 
 			if testName == conf.PerfTestTokenMint.String() && pr.cfg.SkipMintConfirmations {
 				// For minting tests a worker can (if configured) skip waiting for a matching response event
 				// before making itself available for the next job
-				confirmations = 0
+				confirmationsPerAction = 0
 			} else {
-				confirmations = len(pr.nodeURLs)
+				confirmationsPerAction += len(pr.nodeURLs)
 				if testName == conf.PerfTestBlobPrivateMsg.String() || testName == conf.PerfTestPrivateMsg.String() {
-					confirmations = 2
+					confirmationsPerAction += 2
 				}
 			}
 
 			// Wait for worker to confirm the message before proceeding to next task
-			for i := 0; i < confirmations; i++ {
-				select {
-				case <-pr.ctx.Done():
-					return nil
-				case <-pr.wsReceivers[workerID]:
-					break
+
+			for j := 0; j < actionsCompleted; j++ {
+				var nextTrackingID string
+				for i := 0; i < confirmationsPerAction; i++ {
+					select {
+					case <-pr.ctx.Done():
+						return nil
+					case <-pr.wsReceivers[workerID]:
+						break
+					}
 				}
+				nextTrackingID = trackingIDs[0]
+				if len(trackingIDs) > 0 {
+					trackingIDs = trackingIDs[1:]
+				}
+				pr.markTestComplete(nextTrackingID)
 			}
 			log.Infof("%d <-- %s Finished (loop=%d)", workerID, testName, loop)
-			pr.markTestComplete(trackingID)
+
+			// for _, trackingID := range trackingIDs {
```

Comment:
can we just remove this commented out bit?

Reply:
Yes good point, missed that I'd left it in. I've added a new commit to remove.